### DISCUSSION TITLE
Indexing with an array of booleans

### DISCUSSION THREAD

**Discussion by dvd101x** [josdejong/mathjs#2735] [Design decisions]
It would be nice to index with an array of booleans, I'm including some examples and reasoning.

```python
a = [4, 5, 6]
a[[true, false, true]] # I would like for it to return [4, 6]
```
From this I get `IndexError: Index out of range (0 < 1)`

The desired result helps in other languages like Matlab/Octave or Python-Numpy.

## Example in Matlab/Octave

```Matlab
% Let's say I want to filter only even numbers
a = [4, 5, 6]
even = mod(a,2)==0	% returns [1, 0, 1]
a(even)			% returns [4, 6]
```

## Example in Python-Numpy

```python
a = np.array([4, 5, 6])
even = a%2 == 0		# returns array([ True, False,  True])
a[even]			# returns array([4, 6])
```

## Desired example in MathJS

```python
a = [4, 5, 6]
even = a%2 == 0		# returns [true, false, true]
a[even] 		# desired output [4, 6]
```

I understand this might not be an easy task, and also there is a function `filter` for similar issues, nonetheless I bring this topic for considerations as I think It could help extending even more the capabilities of this library.

**Comment by josdejong** [comment]
Thanks, that is indeed a useful way to allow you to filter array items. 

I think it conflicts with the current behavior of `a[indices]` using `indices` to to pick the values at the specified indices, and not a mapping of zeros determining a filter value for every item in the array. I'm not sure how we could support both with this same syntax. Maybe we can come up with a different syntax, something like:

```
a = [4, 5, 6]
even = a%2 == 0		# returns [true, false, true]
filter(a, even)		# desired output [4, 6]
```

**Comment by dvd101x** [comment]
Hi Jos!

Thank you for looking into this! Yes `filter` works very well. I'm gladly surprised by the capabilities of this library

Regarding the use of `filter`, I did some tests by defining functions and using inline functions. I was surprised that it can take any undefined variable as the input for the inline function. `filter([1, 2, 3, 4], myVar>2)  #returns [3, 4]`

In my experience (not that much) is more common for `filter` to use a function as an argument. Here is an example in [Pandas-filter](https://pandas.pydata.org/docs/reference/api/pandas.core.groupby.DataFrameGroupBy.filter.html?highlight=filter#pandas.core.groupby.DataFrameGroupBy.filter) and more common for indexing to be used with an array of booleans, here is also an example in [Pandas-subset to filter](https://pandas.pydata.org/docs/getting_started/intro_tutorials/03_subset_data.html#how-do-i-filter-specific-rows-from-a-dataframe).

So I think `indexing` might benefit from interpreting an array of booleans in some very particular cases like for assigning values.

Let's say that we now want to add 0.1 to the even numbers. We could do it by indexing like so if we already knew what is the index of the even numbers

```python
a = [4, 5, 6]
indexOfEven = [1, 3]
a[indexOfEven] = a[indexOfEven] + 0.1
a						# returns [4.1, 5, 6.1]
```

Or we could get that index of even numbers with `filter`
```python
a = [4, 5, 6]
indexOfEven = filter(1:size(a)[1], a[x]%2 == 0)	# returns [1, 3]
a[indexOfEven] = a[indexOfEven]+0.1
a 						# returns [4.1, 5, 6.1]
```

If it was possible to index with an array of booleans (of the same size) then it could be possible to write

```python
a = [4, 5, 6]
a[a%2 == 0] = a[a%2 == 0] + 0.1
a 						# desired return [4.1, 5, 6.1]
```
Similar to Matlab

```Matlab
a = [4, 5, 6]
a(mod(a,2)==0) += 0.1			% returns [4.1, 5, 6.1]
```

I found a few examples
* [Octave documentation for vectorization](https://docs.octave.org/interpreter/Basic-Vectorization.html)
* [Numpy documentation for indexing](https://numpy.org/doc/stable/user/basics.indexing.html?highlight=slice)

I understand that indexing is already very powerful as each index can be an integer, a vector of integers and a range (including the keyword `end` for the maximum index possible).

So if the index is a vector of the same size and it contains only booleans, then it could apply a conversion to make `[true, false, true, true]` into `[1, 3, 4]` and the rest would be the same as it is.

So we could write something like
```python
mass = [12, 5, 15, 20]
# Get all that masses that are greater than 10
mass[mass>10] # expected return [12, 15, 20]

force = [5, 10, 20, 15]
# Get all the forces where the masses are less than 15
force[mass<15] # expected return [5, 10]
```

**Comment by josdejong** [comment]
Thanks for your inputs. The notation `filter([1, 2, 3, 4], myVar>2)` is just a shorthand for passing an actual function as second argument, like `filter([1, 2, 3, 4], f(myVar) = myVar>2)`.

Interesting idea to extend the index to support functions like `force[mass<15]`. It feels to me like eye-candy for doing `filter(force, mass<15)`, I'm not sure it really adds value. It looks nice though :). I would at least love to extend the `filter` function thought to support a matrix as second argument, like `filter(a, even)`. I still have to give this a a bit more thought.

**Comment by dvd101x** [comment]
Thank you for considering this,

Adding this capability to do `filter(force, mass<15)` sounds very cool too.

As a summary I found references on Octave, Numpy and Pandas to do stuff similar to `force[mass<15]` and `filter(force, mass<15)`.

I just found a nice explanation on Numpy, so here it is. [Numpy Filter Array](https://www.w3schools.com/python/numpy/numpy_array_filter.asp)



Again thanks for considering this

**Comment by gwhitney** [comment]
Seems as though the exact desired API is not quite settled here yet, although it appears that at least allowing `filter` to take a collection of booleans as its second argument is "officially sanctioned" if someone wants to implement it. Hence I will add the design decision label as well, at least until it's ironed out what else if anything might be implemented in response to this issue.

**Reply by dvd101x** [reply]
Hi Glen,

I will try to make an example in the following days.

**Reply by dvd101x** [reply]
Here is an example emulating [pandas: boolean indexing](https://pandas.pydata.org/docs/user_guide/indexing.html#boolean-indexing), I hope this can show it's possible value.

This currently works inside the current expression parser

```python

# function to convert an array of booleans to 1-based index
$(x) = filter(1:size(x)[1], _(n) = x[n]);

s = -3:3

# proposed syntax
# s[s > 0]                            would yield              [1, 2, 3]
s[$(s > 0)]                           # emulated syntax yields [1, 2, 3]

# s[(s < -1) or (s > 0.5)]            would yield              [-3, -2, 1, 2, 3]
s[$((s < -1) or (s > 0.5))]           # emulated syntax yields [-3, -2, 1, 2, 3]

# s[(s < 1) and (s > -2.5)]           would yield              [-2, -1, 0]
s[$((s < 1) and (s > -2.5))]          # emulated syntax yields [-2, -1, 0]

# s[not(s < 0)]                       would yield              [0, 1, 2, 3]
s[$(not(s < 0))]                      # emulated syntax yields [0, 1, 2, 3]

```

# Failed tests

I did some tests importing a modified index function into a new instance of mathjs, I had promising results but I always end up messing with the 1-indexing inside the parser.

My method was creating two functions
- check if an array has only booleans inside
- converting them to an array of numbers

``` javascript
isBooleanIndex = math.typed(
  // checks if all elements in an array are boolean
  {
    'Array': index => index.every(e => math.isBoolean(e)),
    'any': index => false
  }
)

convertBooleanIndex = math.typed(
  {
    // converts [true, false, true, false] to [0,2]
    'Array': (x) => math.filter(math.range(0, x.length), (e, i) => x[i]).toArray()
  }
)
```
And then import a new index function into a new mathjs instance

``` javascript
let mat = math.create()
mat.import({
index: math.typed({
    'Array': (...indices) => math.index(...indices.map(index => isBooleanIndex(index) ? convertBooleanIndex(index) : index))
})
},{override:true})
```
Eeven though it did work to index with an array of booleans, I unintentionally converted the 1-basesd indexing to 0-based inside the parser.

**Reply by josdejong** [reply]
Thanks @dvd101x for the clear examples. Your `$(x)` function is quite nifty :). Working this out for real would indeed require adjusting the build-in `ìndex` function (zero based) and it transform (one based).

As for the specific notation, in https://github.com/josdejong/mathjs/discussions/2735#discussioncomment-3537279 I noted that I think this conflicts with the existing functionality, and the index function cannot see the difference. Do you have any thoughts on that? If there indeed is a conflict, I guess we'll have to look for a different syntax, so instead of `s[(s < -1) or (s > 0.5)]` you could write `filter(s, (x < -1) or (x > 0.5))`, which already works. And like Glen says, we could expand `filter` to accept a collection with booleans as second argument so you can write `filter(s, (s < -1) or (s > 0.5))`.

**Reply by gwhitney** [reply]
Jos, what specifically is the conflict? Seems to me as dvd101x says, the index function can check if its index argument is an Array/Matrix containing only boolean values and is the same shape as the collection being indexed. If so, it could filter. It's true that right now in the expression language if A is [7, 8, 9] then `A[[true, true, true]]` evaluates to [7, 7, 7] because of an automatic conversion from boolean to number somewhere along the way, but that's not a very useful behavior and disrupting that by noticing the entries are boolean and using them to filter instead seems feasible to me... Up to you if it's a desirable change to the operation of mathjs.

**Reply by dvd101x** [reply]
Thanks Jos and Glen,

I'm starting see the error in my attempt, I now know that when I did an override of the index function I lost it's transform inside the parser. I'm getting closer by using a factory function but I'm not there yet and I'm trying to improve my understanding of the extension capabilities.

I agree with Glen in the sense that index automatically converts arguments like `[true, true, true]` to `[1, 1, 1]` and `[true, false, true]` to `[1, 0, 1]` I guess due to the boolean to number conversion. So if internally index did first check if it's an array of only booleans it could convert `[true, true, true]` to `[1, 2, 3]` inside the 1-indexed parser and `[0, 1, 2]` directly on the 0-indexed mathjs.

**Reply by gwhitney** [reply]
Well, I think we should focus on the desired syntax and semantics. Implementation is another thing; I sort of doubt the best way to actually do it internally is to convert the booleans to indices and then call the old indexer. It might be more efficient to detect the all-boolean index and then operate more like filter if so (sharing the code thereof).

**Reply by josdejong** [reply]
Ahh, of course, I had missed that. Yes it makes sense to interpret the input containing booleans only as filter, and numeric values as indexes of the row/column. Ok let's do that 👍 .

An edge case would be if the input is a mix of booleans an numbers. Any preference on how to evaluate that? Maybe best would be to throw an error in that case.

**Reply by gwhitney** [reply]
> input is a mix of booleans an numbers. Any preference on how to evaluate that?

Well, at the moment mathjs has an automatic conversion from boolean to number but not vice versa. That suggests some existing built-in preference for converting a mixed Array/Matrix to numeric indices. I am somewhat of two minds about throwing errors: since the mathjs expression language doesn't really have any way to deal with errors, I think it's good to minimize them. On the other hand, mathjs "silently" doing weird things that might be unexpected by the client is a potential source of undetected bugs. So it's a balancing act. Since 0 indices would likely cause an error anyway, I think in this case it's OK to go with the existing mathjs flow and convert booleans to numbers in the mixed case. That's clearly what's happening right now, so it's the least change to current behavior as well.

**Reply by dvd101x** [reply]
Hi, I did some comparisons for mixed types, only as a reference.

## Numpy/Pandas

Both convert an array of booleans to index only if all elements of the array are booleans. If a single element of the array is not a boolean, then it does the default conversion.

I think it would be like adding an intermediate step to mathjs to first check if the array contains only booleans and then converting it to an array that can be used by `index( )`. Else use the current behavior of mathjs.

## Octave/Matlab

They have a type of matrix called logical, that can only contain booleans, when they do `Vector<5` it returns a *logical* matrix.

When manually making a logical matrix, it takes all the *truthy* inputs as true and *falsey* inputs as false.

I think it would be like implementing a new type in mathjs, and then adding a conversion from that type to a type that can be used by `index( )`.

**Reply by dvd101x** [reply]
An additional comparison:

 **Numpy/Pandas** checks that the size of the array of booleans matches the size of the index and throws an error if not, but Octave doesn't and just throws an error if the index is higher than the available index.

**Reply by josdejong** [reply]
Thanks for your feedbacks and David for doing some research! 

I propose that we go with Glens proposal: when a matrix contains booleans only, we filter the result. With mixed and numeric input we apply the default behavior of interpreting as an index.

**Reply by gwhitney** [reply]
Yes, I think this is just about an actionable proposal at this point, that we could invite PRs for. Here's one remaining edge case that remains to be nailed down, that David alludes to above: Suppse `V = [1,2,3,4]`. What happens when mathjs encounters `V[[true, false, true]]`? The index is an array of all booleans, so it should be filtering. But is this an error because V has four elements but the array only has three elements? Or is it `[1,3]` (unspecified booleans default to `false`)? I'd be fine either way, but it needs to be spelled out for someone to implement this.

Any other loose ends?

**Reply by josdejong** [reply]
Good point. I think in such a case it should throw a dimension mismatch error, similar to say `[1,2,3,4] + [1,2,3]`.

**Reply by dvd101x** [reply]
I agree. Ideally it should throw a dimensional error as Python. If not it would still be an improvement and comparable to Octave.

**Reply by dvd101x** [reply]
Hi, is it ok if I send a PR for this?

**Reply by dvd101x** [reply]
Hi, I included a PR for this #2994

**Reply by josdejong** [reply]
Published now in `v11.10.0`
